File: 프로그래머스/2/12978. 배달/배달.py
from collections import defaultdict

def solution(N, road, K):
    answer, roads = {1:0}, defaultdict(set)
    
    for start, end, time in road:
        roads[start].add((end, time))
        roads[end].add((start, time))
        
    def dfs(curr, total_time):
        if total_time > K:
            return
        for end, time in roads[curr]:
            if end in answer and total_time+time >= answer[end]:   # 재방문 불필요
                continue
            elif total_time + time <= K:    # 더 최단 경로일 경우
                answer[end] = total_time + time
                dfs(end, total_time + time)
    dfs(1, 0)
    return len(answer)

File: 프로그래머스/2/12978. 배달/배달.js
function solution(N, road, K) {
    var answer = 0;
    const roads = {};
    const costs = [...new Array(N)].map((_,i)=>500000)
    
    road.forEach((value)=>{ // 양방향 도로 정보 저장
        let [start,end,cost] = value;
        if (start in roads) roads[start].push([end,cost])
        else roads[start] = [[end,cost]]
        if (end in roads) roads[end].push([start,cost])
        else roads[end] = [[start,cost]]
    })

    function dfs(curr, total_cost){ // 지점 당 최소 거리 계산
        if (total_cost<=costs[curr-1]) costs[curr-1] = total_cost
        else return;

        roads[curr].forEach((value)=>{
            let [end, cost] = value
            if (total_cost + cost <= K) dfs(end, total_cost + cost)
        })
    }
    
    dfs(1,0);
    
    answer += costs.filter(value => value <= K).length;
    return answer;
}